src/types/mappingproxy.rs

```diff
@@ -0,0 +1,597 @@
+
+pyobject_native_type_core!(
+    PyMappingProxy,
+    pyobject_native_static_type_object!(ffi::PyDictProxy_Type),
+    #checkfunction=ffi::PyDictProxy_Check
+);
+
+impl PyMappingProxy {
+    /// Creates a mappingproxy from an object.
+    pub fn new<'py>(py: Python<'py>, elements: &'py PyMapping) -> PyResult<&'py PyMappingProxy> {
+        unsafe {
+            let proxy = ffi::PyDictProxy_New(elements.as_ptr());
+            py.from_owned_ptr_or_err::<PyMappingProxy>(proxy)
+        }
+    }
+
+    /// Returns a new mappingproxy that contains the same key-value pairs as self.
+    ///
+    /// This is equivalent to the Python expression `self.copy()`.
+    pub fn copy(&self) -> PyResult<&PyMapping> {
+        self.call_method0("copy")
+            .and_then(|object| object.downcast::<PyMapping>().map_err(PyErr::from))
+    }
+
+    /// Checks if the mappingproxy is empty, i.e. `len(self) == 0`.
+    pub fn is_empty(&self) -> bool {
+        self.len().unwrap_or_default() == 0
+    }
+
+    /// Gets an item from the mappingproxy.
+    ///    
+    /// Returns `None` if the item is not present, or if an error occurs.
+    ///
+    /// To get a `KeyError` for non-existing keys, use `PyAny::get_item`.
+    pub fn get_item<K>(&self, key: K) -> Option<&PyAny>
+    where
+        K: ToPyObject,
+    {
+        PyAny::get_item(self, key).ok()
+    }
+
+    /// Returns a list of mappingproxy keys.
+    ///
+    /// This is equivalent to the Python expression `list(mappingproxy.keys())`.
+    pub fn keys(&self) -> &PyList {
+        unsafe {
+            PySequence::try_from_unchecked(self.call_method0("keys").unwrap())
+                .to_list()
+                .unwrap()
+        }
+    }
+
+    /// Returns a list of mappingproxy values.
+    ///
+    /// This is equivalent to the Python expression `list(mappingproxy.values())`.
+    pub fn values(&self) -> &PyList {
+        unsafe {
+            PySequence::try_from_unchecked(self.call_method0("values").unwrap())
+                .to_list()
+                .unwrap()
+        }
+    }
+
+    /// Returns a list of mappingproxy items.
+    ///
+    /// This is equivalent to the Python expression `list(mappingproxy.items())`.
+    pub fn items(&self) -> &PyList {
+        unsafe {
+            PySequence::try_from_unchecked(self.call_method0("items").unwrap())
+                .to_list()
+                .unwrap()
+        }
+    }
+
+    ///
+    /// Returns an iterator of `(key, value)` pairs in this mappingproxy.
+    pub fn iter(&self) -> PyMappingProxyIterator<'_> {
+        IntoIterator::into_iter(self)
+    }
+
+    /// Returns `self` cast as a `PyMapping`.
+    pub fn as_mapping(&self) -> &PyMapping {
+        unsafe { PyMapping::try_from_unchecked(self) }
+    }
+}
+
+pub struct PyMappingProxyIterator<'py> {
+    iterator: &'py PyIterator,
+    mappingproxy: &'py PyMappingProxy,
+}
+
+impl<'py> Iterator for PyMappingProxyIterator<'py> {
+    type Item = (&'py PyAny, &'py PyAny);
```

Comment:
I guess iteration here should probably be fallible? Overall I see why you added this and I think it makes sense to iterate pairs given the `PyDictIterator` works the same way.

```suggestion
    type Item = PyResult<(&'py PyAny, &'py PyAny)>;
```